worlds/sc2/client_gui.py

```diff
@@ -216,136 +222,164 @@ def build_mission_table(self, dt) -> None:
         else:
             self.campaign_scroll_panel.border_on = False
         self.last_data_out_of_date = self.ctx.data_out_of_date
-        if not self.ctx.mission_req_table:
+        if len(self.ctx.custom_mission_order) == 0:
             self.campaign_panel.add_widget(Label(text="Connect to a world to see a mission layout here."))
             return
 
+        # if self.ctx.slot_data_version >= 4 and self.ctx.mission_order:
         self.last_checked_locations = self.ctx.checked_locations.copy()
         self.first_check = False
-        self.first_mission = get_first_mission(self.ctx.mission_req_table).mission_name
 
         self.mission_id_to_button = {}
 
-        available_missions, unfinished_missions = calc_unfinished_missions(self.ctx)
+        available_missions, available_layouts, available_campaigns, unfinished_missions = calc_unfinished_nodes(self.ctx)
 
         multi_campaign_layout_height = 0
 
-        for campaign, missions in sorted(self.ctx.mission_req_table.items(), key=lambda item: item[0].id):
-            categories: Dict[str, List[str]] = {}
-
-            # separate missions into categories
-            for mission_index in missions:
-                mission_info = self.ctx.mission_req_table[campaign][mission_index]
-                if mission_info.category not in categories:
-                    categories[mission_info.category] = []
-
-                categories[mission_info.category].append(mission_index)
-
-            max_mission_count = max(len(categories[category]) for category in categories)
-            if max_mission_count == 1:
+        for campaign_idx, campaign in enumerate(self.ctx.custom_mission_order):
+            longest_column = max(len(col) for layout in campaign.layouts for col in layout.missions)
+            if longest_column == 1:
                 campaign_layout_height = 115
             else:
-                campaign_layout_height = (max_mission_count + 2) * 50
+                campaign_layout_height = (longest_column + 2) * 56
```

Comment:
56 is a bit of a magic number here (and so is 2), I think it's related to `MISSION_BUTTON_HEIGHT` below but not sure where the extra 6 is coming from. May be worth pulling that definition up to the top of the function and adding some others so it's clearer how things fit together.